File: facebook-messenger/chats/java/app/src/main/java/ai/codia/x/component/CodiaRoundLayoutUtil.java
package ai.codia.x.component;

import android.graphics.Canvas;
import android.graphics.Outline;
import android.graphics.Path;
import android.graphics.RectF;
import android.view.View;
import android.view.ViewOutlineProvider;

/**
 * Created by codia-figma
 */
public class CodiaRoundLayoutUtil {

    private CodiaRoundLayoutUtil.CanvasRounder canvasRound;
    private final View view;

    public CodiaRoundLayoutUtil(View view) {
        this.view = view;
    }

    public void init(CodiaRoundLayoutUtil.CornersHolder holder) {
        canvasRound = CodiaRoundLayoutUtil.CanvasRounder.create(holder);
        view.setOutlineProvider(CodiaRoundLayoutUtil.RoundOutlineProvider.from(holder));
        view.setClipToOutline(true);
    }

    public void init(int radius) {
        CodiaRoundLayoutUtil.CornersHolder holder = CodiaRoundLayoutUtil.CornersHolder.create(radius);
        init(holder);
    }

    public void clipCanvas(Canvas canvas) {
        canvasRound.clipPath(canvas);
    }

    public void onSizeChanged(int w, int h, int oldw, int oldh) {
        canvasRound.updateSize(w, h);
        view.invalidate();
    }


    /**
     * reset rect radius
     *
     * @param radius
     */
    public void updateRectRadius(float radius) {
        canvasRound.updateRadius(radius);
        view.setOutlineProvider(CodiaRoundLayoutUtil.RoundOutlineProvider.from(radius));
        view.invalidate();
    }

    public void updateRectRadius(float topLeft, float topRight, float bottomRight, float bottomLeft) {
        canvasRound.updateRadius(topLeft, topRight, bottomRight, bottomLeft);
        view.setOutlineProvider(CodiaRoundLayoutUtil.RoundOutlineProvider.from(topLeft, topRight, bottomRight, bottomLeft));
        view.invalidate();
    }

    public static class CornersHolder {
        public final float topLeftRadius;
        public final float topRightRadius;
        public final float bottomRightRadius;
        public final float bottomLeftRadius;

        CornersHolder(float topLeft, float topRight, float bottomRight, float bottomLeft) {
            this.topLeftRadius = topLeft;
            this.topRightRadius = topRight;
            this.bottomRightRadius = bottomRight;
            this.bottomLeftRadius = bottomLeft;
        }

        public static CornersHolder create(float topLeft, float topRight, float bottomRight, float bottomLeft) {
            return new CornersHolder(topLeft, topRight, bottomRight, bottomLeft);
        }

        public static CornersHolder create(float radius) {
            return new CornersHolder(radius, radius, radius, radius);
        }
    }

    static class RoundOutlineProvider extends ViewOutlineProvider {
        final CornersHolder corner;

        RoundOutlineProvider(CornersHolder corner) {
            this.corner = corner;
        }

        @Override
        public void getOutline(View view, Outline outline) {
            final RectF rectF = new RectF(0, 0, view.getMeasuredWidth(), view.getMeasuredHeight());
            final Path path = new Path();
            path.addRoundRect(rectF, new float[]{
                corner.topLeftRadius,
                corner.topLeftRadius,
                corner.topRightRadius,
                corner.topRightRadius,
                corner.bottomRightRadius,
                corner.bottomRightRadius,
                corner.bottomLeftRadius,
                corner.bottomLeftRadius,
            }, Path.Direction.CW);
            path.close();
            outline.setConvexPath(path);
        }

        static RoundOutlineProvider from(CornersHolder holder) {
            return new RoundOutlineProvider(holder);
        }

        static RoundOutlineProvider from(float radius) {
            return new RoundOutlineProvider(CornersHolder.create(radius));
        }

        static RoundOutlineProvider from(float topLeft, float topRight, float bottomRight, float bottomLeft) {
            return new RoundOutlineProvider(CornersHolder.create(topLeft, topRight, bottomRight, bottomLeft));
        }
    }

    abstract static class CanvasRounder {
        abstract void clipPath(Canvas canvas);

        abstract void updateSize(int w, int h);

        abstract void updateRadius(float radius);

        abstract void updateRadius(float topLeft, float topRight, float bottomRight, float bottomLeft);

        public static CanvasRounder create(CornersHolder corner) {
            return new RealCanvasRounder(corner);
        }
    }

    static class RealCanvasRounder extends CanvasRounder {
        private final Path path = new Path();
        private CornersHolder corner;
        private final RectF rectF = new RectF(0f, 0f, 0f, 0f);

        RealCanvasRounder(CornersHolder corner) {
            this.corner = corner;
        }

        void clipPath(Canvas canvas) {
            canvas.clipPath(path);
        }

        void updateSize(int w, int h) {
            rectF.set(0, 0, w, h);
            updateRadius();
        }

        @Override
        void updateRadius(float radius) {
            corner = CornersHolder.create(radius);
            updateRadius();
        }

        @Override
        void updateRadius(float topLeft, float topRight, float bottomRight, float bottomLeft) {
            corner = CornersHolder.create(topLeft, topRight, bottomRight, bottomLeft);
            updateRadius();
        }

        private void updateRadius() {
            path.reset();
            path.addRoundRect(rectF, new float[]{
                    corner.topLeftRadius,
                    corner.topLeftRadius,
                    corner.topRightRadius,
                    corner.topRightRadius,
                    corner.bottomRightRadius,
                    corner.bottomRightRadius,
                    corner.bottomLeftRadius,
                    corner.bottomLeftRadius,
            }, Path.Direction.CW);
            path.close();
        }
    }
}

File: facebook-messenger/discover/kotlin/app/src/main/java/ai/codia/x/component/CodiaRoundLayoutUtil.kt
package ai.codia.x.component

import android.graphics.Canvas
import android.graphics.Outline
import android.graphics.Path
import android.graphics.RectF
import android.view.View
import android.view.ViewOutlineProvider

/**
 * Created by codia-figma
 */
class CodiaRoundLayoutUtil(private val view: View) {
    private var canvasRounder: CanvasRounder? = null
    fun init(holder: CornersHolder) {
        canvasRounder = CanvasRounder.create(holder)
        view.outlineProvider = RoundOutlineProvider.from(holder)
        view.setClipToOutline(true)
    }

    fun init(radius: Int) {
        val holder = CornersHolder.create(radius.toFloat())
        init(holder)
    }

    fun clipCanvas(canvas: Canvas) {
        canvasRounder!!.clipPath(canvas)
    }

    fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        canvasRounder!!.updateSize(w, h)
        view.invalidate()
    }

    /**
     * reset rect radius
     *
     * @param radius
     */
    fun updateRectRadius(radius: Float) {
        canvasRounder!!.updateRadius(radius)
        view.outlineProvider = RoundOutlineProvider.from(radius)
        view.invalidate()
    }

    fun updateRectRadius(topLeft: Float, topRight: Float, bottomRight: Float, bottomLeft: Float) {
        canvasRounder!!.updateRadius(topLeft, topRight, bottomRight, bottomLeft)
        view.outlineProvider = RoundOutlineProvider.from(topLeft, topRight, bottomRight, bottomLeft)
        view.invalidate()
    }

    class CornersHolder internal constructor(
        val topLeftRadius: Float,
        val topRightRadius: Float,
        val bottomRightRadius: Float,
        val bottomLeftRadius: Float
    ) {
        companion object {
            fun create(
                topLeft: Float, topRight: Float, bottomRight: Float, bottomLeft: Float
            ): CornersHolder {
                return CornersHolder(topLeft, topRight, bottomRight, bottomLeft)
            }

            fun create(radius: Float): CornersHolder {
                return CornersHolder(radius, radius, radius, radius)
            }
        }
    }

    internal class RoundOutlineProvider(val corner: CornersHolder) : ViewOutlineProvider() {
        override fun getOutline(view: View, outline: Outline) {
            val rectF = RectF(0f, 0f, view.measuredWidth.toFloat(), view.measuredHeight.toFloat())
            val path = Path()
            path.addRoundRect(
                rectF, floatArrayOf(
                    corner.topLeftRadius,
                    corner.topLeftRadius,
                    corner.topRightRadius,
                    corner.topRightRadius,
                    corner.bottomRightRadius,
                    corner.bottomRightRadius,
                    corner.bottomLeftRadius,
                    corner.bottomLeftRadius
                ), Path.Direction.CW
            )
            path.close()
            outline.setConvexPath(path)
        }

        companion object {
            fun from(holder: CornersHolder): RoundOutlineProvider {
                return RoundOutlineProvider(holder)
            }

            fun from(radius: Float): RoundOutlineProvider {
                return RoundOutlineProvider(CornersHolder.create(radius))
            }

            fun from(
                topLeft: Float, topRight: Float, bottomRight: Float, bottomLeft: Float
            ): RoundOutlineProvider {
                return RoundOutlineProvider(
                    CornersHolder.create(
                        topLeft, topRight, bottomRight, bottomLeft
                    )
                )
            }
        }
    }

    internal abstract class CanvasRounder {
        abstract fun clipPath(canvas: Canvas)
        abstract fun updateSize(w: Int, h: Int)
        abstract fun updateRadius(radius: Float)
        abstract fun updateRadius(
            topLeft: Float, topRight: Float, bottomRight: Float, bottomLeft: Float
        )

        companion object {
            fun create(corner: CornersHolder): CanvasRounder {
                return RealCanvasRounder(corner)
            }
        }
    }

    internal class RealCanvasRounder(private var corner: CornersHolder) : CanvasRounder() {
        private val path = Path()
        private val rectF = RectF(0f, 0f, 0f, 0f)
        override fun clipPath(canvas: Canvas) {
            canvas.clipPath(path)
        }

        override fun updateSize(w: Int, h: Int) {
            rectF[0f, 0f, w.toFloat()] = h.toFloat()
            updateRadius()
        }

        override fun updateRadius(radius: Float) {
            corner = CornersHolder.create(radius)
            updateRadius()
        }

        override fun updateRadius(
            topLeft: Float, topRight: Float, bottomRight: Float, bottomLeft: Float
        ) {
            corner = CornersHolder.create(topLeft, topRight, bottomRight, bottomLeft)
            updateRadius()
        }

        private fun updateRadius() {
            path.reset()
            path.addRoundRect(
                rectF, floatArrayOf(
                    corner.topLeftRadius,
                    corner.topLeftRadius,
                    corner.topRightRadius,
                    corner.topRightRadius,
                    corner.bottomRightRadius,
                    corner.bottomRightRadius,
                    corner.bottomLeftRadius,
                    corner.bottomLeftRadius
                ), Path.Direction.CW
            )
            path.close()
        }
    }
}
